File: src/Crawler/Result/VisitedUrl.php
<?php

/*
 * This file is part of the SiteOne Crawler.
 *
 * (c) Ján Regeš <jan.reges@siteone.cz>
 */

declare(strict_types=1);

namespace Crawler\Result;

use Crawler\Crawler;
use Crawler\FoundUrl;
use Crawler\Utils;

class VisitedUrl
{
    const ERROR_CONNECTION_FAIL = -1;
    const ERROR_TIMEOUT = -2;
    const ERROR_SERVER_RESET = -3;
    const ERROR_SEND_ERROR = -4;
    const ERROR_SKIPPED = -6;

    // cache type flags for the cacheType property (uses bitwise OR)
    const CACHE_TYPE_HAS_CACHE_CONTROL = 1;
    const CACHE_TYPE_HAS_EXPIRES = 2;
    const CACHE_TYPE_HAS_ETAG = 4;
    const CACHE_TYPE_HAS_LAST_MODIFIED = 8;
    const CACHE_TYPE_HAS_MAX_AGE = 16;
    const CACHE_TYPE_HAS_S_MAX_AGE = 32;
    const CACHE_TYPE_HAS_STALE_WHILE_REVALIDATE = 64;
    const CACHE_TYPE_HAS_STALE_IF_ERROR = 128;
    const CACHE_TYPE_HAS_PUBLIC = 256;
    const CACHE_TYPE_HAS_PRIVATE = 512;
    const CACHE_TYPE_HAS_NO_CACHE = 1024;
    const CACHE_TYPE_HAS_NO_STORE = 2048;
    const CACHE_TYPE_HAS_MUST_REVALIDATE = 4096;
    const CACHE_TYPE_HAS_PROXY_REVALIDATE = 8192;
    const CACHE_TYPE_HAS_IMMUTABLE = 16384;
    const CACHE_TYPE_NO_CACHE_HEADERS = 32768;
    const CACHE_TYPE_NOT_AVAILABLE = 65536;

    /**
     * @var string Unique ID hash of this URL
     */
    public readonly string $uqId;

    /**
     * @var string Unique ID hash of the source URL where this URL was found
     */
    public readonly string $sourceUqId;

    /**
     * @var int Source attribute where this URL was found. See FoundUrl::SOURCE_* constants
     */
    public readonly int $sourceAttr;

    /**
     * Full URL with scheme, domain, path and query
     * @var string URL
     */
    public readonly string $url;

    /**
     * HTTP status code of the request
     * Negative values are errors - see self:ERROR_* constants
     * @var int
     */
    public readonly int $statusCode;

    /**
     * Request time in seconds
     * @var float
     */
    public readonly float $requestTime;

    /**
     * Request time formatted as "32 ms" or "7.4 s"
     * @var string
     */
    public readonly string $requestTimeFormatted;

    /**
     * Size of the response in bytes
     * @var int|null
     */
    public readonly ?int $size;

    /**
     * Size of the response formatted as "1.23 MB"
     * @var string|null
     */
    public readonly ?string $sizeFormatted;

    /**
     * Content-Encoding header value (br, gzip, ...)
     * @var string|null
     */
    public readonly ?string $contentEncoding;

    /**
     * Content type ID
     * @see Crawler::CONTENT_TYPE_ID_*
     * @var int
     */
    public readonly int $contentType;

    /**
     * Content type header value (text/html, application/json, ...)
     * @var string|null
     */
    public readonly ?string $contentTypeHeader;

    /**
     * Extra data from the response required by --extra-columns (headers, Title, DOM, etc.
     * @var array|null
     */
    public readonly ?array $extras;

    /**
     * Is this URL external (not from the same domain as the initial URL)
     * @var bool
     */
    public readonly bool $isExternal;

    /**
     * Is this URL allowed for crawling (based on --allowed-domain-for-crawling)
     * @var bool
     */
    public readonly bool $isAllowedForCrawling;

    /**
     * Cache type flags of the response (bitwise OR). See self::CACHE_TYPE_* constants
     * @var int
     */
    public readonly int $cacheTypeFlags;

    /**
     * How long the response is allowed to be cached in seconds (based on max-age in Cache-Control or Expires header)
     * @var int|null
     */
    public readonly int|null $cacheLifetime;

    /**
     * @var string|null
     */
    private ?string $hostCache = null;

    /**
     * @param string $uqId
     * @param string $sourceUqId
     * @param int $sourceAttr
     * @param string $url
     * @param int $statusCode
     * @param float $requestTime
     * @param int|null $size
     * @param int $contentType
     * @param string|null $contentTypeHeader
     * @param string|null $contentEncoding
     * @param array|null $extras
     * @param bool $isExternal
     * @param bool $isAllowedForCrawling
     * @param int $cacheType
     * @param int|null $cacheLifetime
     */
    public function __construct(string $uqId, string $sourceUqId, int $sourceAttr, string $url, int $statusCode, float $requestTime, ?int $size, int $contentType, ?string $contentTypeHeader, ?string $contentEncoding, ?array $extras, bool $isExternal, bool $isAllowedForCrawling, int $cacheType, ?int $cacheLifetime)
    {
        $this->uqId = $uqId;
        $this->sourceUqId = $sourceUqId;
        $this->sourceAttr = $sourceAttr;
        $this->url = $url;
        $this->statusCode = $statusCode;
        $this->requestTime = $requestTime;
        $this->requestTimeFormatted = Utils::getFormattedDuration($this->requestTime);
        $this->size = $size;
        $this->sizeFormatted = $size !== null ? Utils::getFormattedSize($size) : null;
        $this->contentType = $contentType;
        $this->contentTypeHeader = $contentTypeHeader;
        $this->contentEncoding = $contentEncoding;
        $this->extras = $extras;
        $this->isExternal = $isExternal;
        $this->isAllowedForCrawling = $isAllowedForCrawling;
        $this->cacheTypeFlags = $cacheType;
        $this->cacheLifetime = $cacheLifetime;
    }

    public function isHttps(): bool
    {
        return str_starts_with($this->url, 'https://');
    }

    public function isStaticFile(): bool
    {
        static $staticTypes = [
            Crawler::CONTENT_TYPE_ID_IMAGE,
            Crawler::CONTENT_TYPE_ID_SCRIPT,
            Crawler::CONTENT_TYPE_ID_STYLESHEET,
            Crawler::CONTENT_TYPE_ID_VIDEO,
            Crawler::CONTENT_TYPE_ID_AUDIO,
            Crawler::CONTENT_TYPE_ID_DOCUMENT,
            Crawler::CONTENT_TYPE_ID_FONT,
            Crawler::CONTENT_TYPE_ID_JSON,
            Crawler::CONTENT_TYPE_ID_XML,
        ];

        return in_array($this->contentType, $staticTypes);
    }

    public function isImage(): bool
    {
        return $this->contentType === Crawler::CONTENT_TYPE_ID_IMAGE;
    }

    public function isVideo(): bool
    {
        return $this->contentType === Crawler::CONTENT_TYPE_ID_VIDEO;
    }

    /**
     * @param string|null $sourceUrl
     * @return string
     */
    public function getSourceDescription(?string $sourceUrl): string
    {
        return match ($this->sourceAttr) {
            FoundUrl::SOURCE_INIT_URL => 'Initial URL',
            FoundUrl::SOURCE_A_HREF => "<a href> on $sourceUrl",
            FoundUrl::SOURCE_IMG_SRC => "<img src> on $sourceUrl",
            FoundUrl::SOURCE_IMG_SRCSET => "<img srcset> on $sourceUrl",
            FoundUrl::SOURCE_INPUT_SRC => "<input src> on $sourceUrl",
            FoundUrl::SOURCE_SOURCE_SRC => "<source src> on $sourceUrl",
            FoundUrl::SOURCE_VIDEO_SRC => "<video src> on $sourceUrl",
            FoundUrl::SOURCE_AUDIO_SRC => "<audio src> on $sourceUrl",
            FoundUrl::SOURCE_SCRIPT_SRC => "<script src> on $sourceUrl",
            FoundUrl::SOURCE_INLINE_SCRIPT_SRC => "<script> on $sourceUrl",
            FoundUrl::SOURCE_LINK_HREF => "<link href> on $sourceUrl",
            FoundUrl::SOURCE_CSS_URL => "CSS url() on $sourceUrl",
            FoundUrl::SOURCE_JS_URL => "JS url on $sourceUrl",
            FoundUrl::SOURCE_REDIRECT => "Redirect from $sourceUrl",
            FoundUrl::SOURCE_SITEMAP => "URL in sitemap $sourceUrl",
            default => 'Unknown source',
        };
    }

    /**
     * @return string
     */
    public function getSourceShortName(): string
    {
        return match ($this->sourceAttr) {
            FoundUrl::SOURCE_INIT_URL => 'Initial URL',
            FoundUrl::SOURCE_A_HREF => '<a href>',
            FoundUrl::SOURCE_IMG_SRC => '<img src>',
            FoundUrl::SOURCE_IMG_SRCSET => '<img srcset>',
            FoundUrl::SOURCE_INPUT_SRC => '<input src>',
            FoundUrl::SOURCE_SOURCE_SRC => '<source src>',
            FoundUrl::SOURCE_VIDEO_SRC => '<video src>',
            FoundUrl::SOURCE_AUDIO_SRC => '<audio src>',
            FoundUrl::SOURCE_SCRIPT_SRC => '<script src>',
            FoundUrl::SOURCE_INLINE_SCRIPT_SRC => 'inline <script src>',
            FoundUrl::SOURCE_LINK_HREF => '<link href>',
            FoundUrl::SOURCE_CSS_URL => 'css url()',
            FoundUrl::SOURCE_JS_URL => 'js url',
            FoundUrl::SOURCE_REDIRECT => 'redirect',
            FoundUrl::SOURCE_SITEMAP => 'sitemap',
            default => 'unknown',
        };
    }

    public function looksLikeStaticFileByUrl(): bool
    {
        return preg_match('/\.(jpg|jpeg|png|gif|webp|svg|ico|js|css|txt|woff2|woff|ttf|eot|mp4|webm|ogg|mp3|wav|flac|pdf|doc|docx|xls|xlsx|ppt|pptx|zip|rar|gz|bz2|7z|xml|json)/i', $this->url) === 1;
    }

    public function hasErrorStatusCode(): bool
    {
        return $this->statusCode < 0;
    }

    public function getScheme(): string
    {
        return parse_url($this->url, PHP_URL_SCHEME);
    }

    public function getHost(): string
    {
        if ($this->hostCache !== null) {
            return $this->hostCache;
        }
        $this->hostCache = parse_url($this->url, PHP_URL_HOST);
        return $this->hostCache;
    }

    public function getPort(): int
    {
        $port = parse_url($this->url, PHP_URL_PORT);
        if ($port === null) {
            $port = $this->isHttps() ? 443 : 80;
        }
        return (int)$port;
    }

    public function getCacheTypeLabel(): string
    {
        $labels = [];

        // Cache-Control or Expires (if Cache-Control is not defined)
        if ($this->cacheTypeFlags & self::CACHE_TYPE_HAS_CACHE_CONTROL) {
            $labels[] = 'Cache-Control';
        } else if ($this->cacheTypeFlags & self::CACHE_TYPE_HAS_EXPIRES) {
            $labels[] = 'Expires';
        }

        // ETag and Last-Modified
        if ($this->cacheTypeFlags & self::CACHE_TYPE_HAS_ETAG) {
            $labels[] = 'ETag';
        }
        if ($this->cacheTypeFlags & self::CACHE_TYPE_HAS_LAST_MODIFIED) {
            $labels[] = 'Last-Modified';
        }

        return $labels ? implode(' + ', $labels) : 'No cache headers';
    }

}

File: src/result/visited_url.rs
// SiteOne Crawler - VisitedUrl
// (c) Jan Reges <jan.reges@siteone.cz>

use std::collections::HashMap;

use regex::Regex;
use serde::{Deserialize, Serialize};

use crate::types::ContentTypeId;
use crate::utils;

// Error status codes (negative values)
pub const ERROR_CONNECTION_FAIL: i32 = -1;
pub const ERROR_TIMEOUT: i32 = -2;
pub const ERROR_SERVER_RESET: i32 = -3;
pub const ERROR_SEND_ERROR: i32 = -4;
pub const ERROR_SKIPPED: i32 = -6;

// Cache type flags (bitwise OR)
pub const CACHE_TYPE_HAS_CACHE_CONTROL: u32 = 1;
pub const CACHE_TYPE_HAS_EXPIRES: u32 = 2;
pub const CACHE_TYPE_HAS_ETAG: u32 = 4;
pub const CACHE_TYPE_HAS_LAST_MODIFIED: u32 = 8;
pub const CACHE_TYPE_HAS_MAX_AGE: u32 = 16;
pub const CACHE_TYPE_HAS_S_MAX_AGE: u32 = 32;
pub const CACHE_TYPE_HAS_STALE_WHILE_REVALIDATE: u32 = 64;
pub const CACHE_TYPE_HAS_STALE_IF_ERROR: u32 = 128;
pub const CACHE_TYPE_HAS_PUBLIC: u32 = 256;
pub const CACHE_TYPE_HAS_PRIVATE: u32 = 512;
pub const CACHE_TYPE_HAS_NO_CACHE: u32 = 1024;
pub const CACHE_TYPE_HAS_NO_STORE: u32 = 2048;
pub const CACHE_TYPE_HAS_MUST_REVALIDATE: u32 = 4096;
pub const CACHE_TYPE_HAS_PROXY_REVALIDATE: u32 = 8192;
pub const CACHE_TYPE_HAS_IMMUTABLE: u32 = 16384;
pub const CACHE_TYPE_NO_CACHE_HEADERS: u32 = 32768;
pub const CACHE_TYPE_NOT_AVAILABLE: u32 = 65536;

// Source attribute constants
pub const SOURCE_INIT_URL: i32 = 5;
pub const SOURCE_A_HREF: i32 = 10;
pub const SOURCE_IMG_SRC: i32 = 20;
pub const SOURCE_IMG_SRCSET: i32 = 21;
pub const SOURCE_INPUT_SRC: i32 = 22;
pub const SOURCE_SOURCE_SRC: i32 = 23;
pub const SOURCE_VIDEO_SRC: i32 = 24;
pub const SOURCE_AUDIO_SRC: i32 = 25;
pub const SOURCE_SCRIPT_SRC: i32 = 30;
pub const SOURCE_INLINE_SCRIPT_SRC: i32 = 40;
pub const SOURCE_LINK_HREF: i32 = 50;
pub const SOURCE_CSS_URL: i32 = 60;
pub const SOURCE_JS_URL: i32 = 70;
pub const SOURCE_REDIRECT: i32 = 80;
pub const SOURCE_SITEMAP: i32 = 90;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VisitedUrl {
    /// Unique ID hash of this URL
    pub uq_id: String,

    /// Unique ID hash of the source URL where this URL was found
    pub source_uq_id: String,

    /// Source attribute where this URL was found (see SOURCE_* constants)
    pub source_attr: i32,

    /// Full URL with scheme, domain, path and query
    pub url: String,

    /// HTTP status code of the request (negative values are errors, see ERROR_* constants)
    pub status_code: i32,

    /// Request time in seconds
    pub request_time: f64,

    /// Request time formatted as "32 ms" or "7.4 s"
    pub request_time_formatted: String,

    /// Size of the response in bytes
    pub size: Option<i64>,

    /// Size of the response formatted as "1.23 MB"
    pub size_formatted: Option<String>,

    /// Content-Encoding header value (br, gzip, ...)
    pub content_encoding: Option<String>,

    /// Content type ID
    pub content_type: ContentTypeId,

    /// Content type header value (text/html, application/json, ...)
    pub content_type_header: Option<String>,

    /// Extra data from the response required by --extra-columns
    pub extras: Option<HashMap<String, String>>,

    /// Is this URL external (not from the same domain as the initial URL)
    pub is_external: bool,

    /// Is this URL allowed for crawling (based on --allowed-domain-for-crawling)
    pub is_allowed_for_crawling: bool,

    /// Cache type flags of the response (bitwise OR). See CACHE_TYPE_* constants
    pub cache_type_flags: u32,

    /// How long the response is allowed to be cached in seconds
    pub cache_lifetime: Option<i64>,
}

impl VisitedUrl {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        uq_id: String,
        source_uq_id: String,
        source_attr: i32,
        url: String,
        status_code: i32,
        request_time: f64,
        size: Option<i64>,
        content_type: ContentTypeId,
        content_type_header: Option<String>,
        content_encoding: Option<String>,
        extras: Option<HashMap<String, String>>,
        is_external: bool,
        is_allowed_for_crawling: bool,
        cache_type_flags: u32,
        cache_lifetime: Option<i64>,
    ) -> Self {
        let request_time_formatted = utils::get_formatted_duration(request_time);
        let size_formatted = size.map(|s| utils::get_formatted_size(s, 0));

        Self {
            uq_id,
            source_uq_id,
            source_attr,
            url,
            status_code,
            request_time,
            request_time_formatted,
            size,
            size_formatted,
            content_encoding,
            content_type,
            content_type_header,
            extras,
            is_external,
            is_allowed_for_crawling,
            cache_type_flags,
            cache_lifetime,
        }
    }

    pub fn is_https(&self) -> bool {
        self.url.starts_with("https://")
    }

    pub fn is_static_file(&self) -> bool {
        matches!(
            self.content_type,
            ContentTypeId::Image
                | ContentTypeId::Script
                | ContentTypeId::Stylesheet
                | ContentTypeId::Video
                | ContentTypeId::Audio
                | ContentTypeId::Document
                | ContentTypeId::Font
                | ContentTypeId::Json
                | ContentTypeId::Xml
        )
    }

    pub fn is_image(&self) -> bool {
        self.content_type == ContentTypeId::Image
    }

    pub fn is_video(&self) -> bool {
        self.content_type == ContentTypeId::Video
    }

    pub fn get_source_description(&self, source_url: Option<&str>) -> String {
        let source = source_url.unwrap_or("unknown");
        match self.source_attr {
            SOURCE_INIT_URL => "Initial URL".to_string(),
            SOURCE_A_HREF => format!("<a href> on {}", source),
            SOURCE_IMG_SRC => format!("<img src> on {}", source),
            SOURCE_IMG_SRCSET => format!("<img srcset> on {}", source),
            SOURCE_INPUT_SRC => format!("<input src> on {}", source),
            SOURCE_SOURCE_SRC => format!("<source src> on {}", source),
            SOURCE_VIDEO_SRC => format!("<video src> on {}", source),
            SOURCE_AUDIO_SRC => format!("<audio src> on {}", source),
            SOURCE_SCRIPT_SRC => format!("<script src> on {}", source),
            SOURCE_INLINE_SCRIPT_SRC => format!("<script> on {}", source),
            SOURCE_LINK_HREF => format!("<link href> on {}", source),
            SOURCE_CSS_URL => format!("CSS url() on {}", source),
            SOURCE_JS_URL => format!("JS url on {}", source),
            SOURCE_REDIRECT => format!("Redirect from {}", source),
            SOURCE_SITEMAP => format!("URL in sitemap {}", source),
            _ => "Unknown source".to_string(),
        }
    }

    pub fn get_source_short_name(&self) -> &'static str {
        match self.source_attr {
            SOURCE_INIT_URL => "Initial URL",
            SOURCE_A_HREF => "<a href>",
            SOURCE_IMG_SRC => "<img src>",
            SOURCE_IMG_SRCSET => "<img srcset>",
            SOURCE_INPUT_SRC => "<input src>",
            SOURCE_SOURCE_SRC => "<source src>",
            SOURCE_VIDEO_SRC => "<video src>",
            SOURCE_AUDIO_SRC => "<audio src>",
            SOURCE_SCRIPT_SRC => "<script src>",
            SOURCE_INLINE_SCRIPT_SRC => "inline <script src>",
            SOURCE_LINK_HREF => "<link href>",
            SOURCE_CSS_URL => "css url()",
            SOURCE_JS_URL => "js url",
            SOURCE_REDIRECT => "redirect",
            SOURCE_SITEMAP => "sitemap",
            _ => "unknown",
        }
    }

    pub fn looks_like_static_file_by_url(&self) -> bool {
        use once_cell::sync::Lazy;
        static RE_STATIC_FILE: Lazy<Regex> = Lazy::new(|| {
            Regex::new(
                r"(?i)\.(jpg|jpeg|png|gif|webp|svg|ico|js|css|txt|woff2|woff|ttf|eot|mp4|webm|ogg|mp3|wav|flac|pdf|doc|docx|xls|xlsx|ppt|pptx|zip|rar|gz|bz2|7z|xml|json)",
            ).unwrap()
        });
        RE_STATIC_FILE.is_match(&self.url)
    }

    pub fn has_error_status_code(&self) -> bool {
        self.status_code < 0
    }

    pub fn get_scheme(&self) -> Option<String> {
        url::Url::parse(&self.url).ok().map(|u| u.scheme().to_string())
    }

    pub fn get_host(&self) -> Option<String> {
        url::Url::parse(&self.url)
            .ok()
            .and_then(|u| u.host_str().map(|h| h.to_string()))
    }

    pub fn get_port(&self) -> u16 {
        if let Ok(parsed) = url::Url::parse(&self.url) {
            parsed.port().unwrap_or_else(|| if self.is_https() { 443 } else { 80 })
        } else if self.is_https() {
            443
        } else {
            80
        }
    }

    pub fn get_cache_type_label(&self) -> String {
        let mut labels = Vec::new();

        // Cache-Control or Expires (if Cache-Control is not defined)
        if self.cache_type_flags & CACHE_TYPE_HAS_CACHE_CONTROL != 0 {
            labels.push("Cache-Control");
        } else if self.cache_type_flags & CACHE_TYPE_HAS_EXPIRES != 0 {
            labels.push("Expires");
        }

        // ETag and Last-Modified
        if self.cache_type_flags & CACHE_TYPE_HAS_ETAG != 0 {
            labels.push("ETag");
        }
        if self.cache_type_flags & CACHE_TYPE_HAS_LAST_MODIFIED != 0 {
            labels.push("Last-Modified");
        }

        if labels.is_empty() {
            "No cache headers".to_string()
        } else {
            labels.join(" + ")
        }
    }
}
